### PR TITLE
Bitmap command interface refactor

### PR DESCRIPTION
**Pull request by alexr-bq**
### Summary

Refactors to properly split Bitmap interfaces between glide and SER-compat interfaces. Also fixes overflow issue and adds missing SER-compat interface overload.

### Issue Link

This pull request is linked to issue: [270](https://github.com/valkey-io/valkey-glide-csharp/issues/270)

### Features and Behaviour Changes

This PR splits bitmap support between SER-compatible interfaces and interfaces matching general glide naming

- **GETBIT** (`GetBitAsync`, `StringGetBitAsync`) - Returns the bit value at offset in a string
- **SETBIT** (`SetBitAsync`, `StringSetBitAsync`) - Sets or clears the bit at offset in a string
- **BITCOUNT** (`BitCountAsync`, `StringBitCountAsync`) - Counts set bits in a string with optional byte/bit range
- **BITPOS** (`BitPosAsync`, `StringBitPositionAsync`) - Finds first bit set to 0 or 1 with optional range
- **BITOP** (`BitOpAsync`, `StringBitOperationAsync`) - Performs bitwise AND, OR, XOR, NOT operations between keys
- **BITFIELD** (`BitFieldAsync`) - Reads/modifies arbitrary bit fields with GET, SET, INCRBY subcommands and overflow control
- **BITFIELD_RO** (`BitFieldReadOnlyAsync`) - Read-only variant of BITFIELD that can be routed to replicas

Two API styles are provided:
- **GLIDE-style** (`IBaseClient`): `GetBitAsync`, `SetBitAsync`, `BitCountAsync`, `BitPosAsync`, `BitOpAsync`, `BitFieldAsync`, `BitFieldReadOnlyAsync`
- **StackExchange.Redis-compatible** (`IDatabaseAsync`): `StringGetBitAsync`, `StringSetBitAsync`, `StringBitCountAsync`, `StringBitPositionAsync`, `StringBitOperationAsync`

### Implementation

Key implementation details:

- **Interface separation**: GLIDE-style methods in `IBaseClient.BitmapCommands.cs`, SER-compatible methods in `IDatabaseAsync.BitmapCommands.cs`. `IBitmapBaseCommands` is intentionally empty as there's no naming overlap between the two styles.
- **Index type enums**: `BitmapIndexType` for GLIDE methods, `StringIndexType` for SER-compatible methods (both support Byte/Bit indexing)
- **BitFieldOptions**: Comprehensive subcommand support including `Get`, `Set`, `Incrby`, and `Overflow` with WRAP/SAT/FAIL modes
- **BitOffsetOptions**: Encapsulates start/end/indexType for cleaner overloads
- **Request building**: `Request.BitmapCommands.cs` handles argument construction for all bitmap commands
- **Response parsing**: `ResponseParser.cs` extended with `ParseBitFieldResponse` for nullable long arrays (OVERFLOW FAIL returns null)

Areas for reviewer attention:
- `BitFieldOptions.cs` - Complex subcommand hierarchy with signed/unsigned field encoding
- `BaseClient.BitmapCommands.cs` - Core implementation delegating to request builders
- `Database.BitmapCommands.cs` - SER-compatible wrappers with CommandFlags validation

### Limitations

- **CommandFlags**: SER-compatible methods accept `CommandFlags` parameter but only `CommandFlags.None` is supported. Non-None flags throw `NotImplementedException`.
- **BITFIELD/BITFIELD_RO**: Only available through GLIDE-style API (`IBaseClient`), not SER-compatible interface, as StackExchange.Redis doesn't support these commands.

### Testing

- 184 bitmap-related integration tests covering:
  - All bitmap commands (GETBIT, SETBIT, BITCOUNT, BITPOS, BITOP, BITFIELD, BITFIELD_RO)
  - Both standalone and cluster modes
  - Edge cases: non-existent keys, offset beyond string length, empty strings
  - BITFIELD overflow modes (WRAP, SAT, FAIL) with null handling for FAIL mode
  - SER compatibility tests verifying CommandFlags.None acceptance and non-None rejection
  - API equivalence tests comparing GLIDE methods to SER counterparts

All tests passing

### Checklist

- [ ] This Pull Request is related to one issue.
- [ ] Commit message has a detailed description of what changed and why.
- [ ] Tests are added or updated and all checks pass.
- [ ] `CHANGELOG.md`, `README.md`, `DEVELOPER.md`, and other documentation files are updated.
- [ ] Destination branch is correct - `main` or release
- [ ] Create merge commit if merging release branch into `main`, squash otherwise.

### CODE REVIEW

File: sources/Valkey.Glide/Abstract/IDatabaseAsync.BitmapCommands.cs [hunk 1]
```diff
@@ -1,53 +1,97 @@
...
+    /// Returns the bit value at offset in the string value stored at key.
+    /// </summary>
+    /// <seealso href="https://valkey.io/commands/getbit"/>
+    /// <param name="key">The key of the string.</param>
+    /// <param name="offset">The offset in the string to get the bit at.</param>
     /// <param name="flags">Command flags (currently not supported by GLIDE).</param>
+    /// <returns>The bit value stored at offset. Returns false if the key does not exist or if the offset is beyond the string length.</returns>
     /// <exception cref="NotImplementedException">Thrown if <paramref name="flags"/> is not <see cref="CommandFlags.None"/>.</exception>
-    Task<bool> StringGetBitAsync(ValkeyKey key, long offset, CommandFlags flags);
+    Task<bool> StringGetBitAsync(ValkeyKey key, long offset, CommandFlags flags = CommandFlags.None);
```
Comment (currantw) [inline sources/Valkey.Glide/Abstract/IDatabaseAsync.BitmapCommands.cs]: Can we inherit and then override the GLIDE documentation? I think we do that in many other spots already? (Again, feel free not to address for time concerns).

File: sources/Valkey.Glide/Internals/Request.BitmapCommands.cs [hunk 2]
```diff
@@ -14,48 +14,58 @@ public static Cmd<long, bool> GetBitAsync(ValkeyKey key, long offset)
...
         {
             args.Add(indexType.ToLiteral().ToGlideString());
         }
         return Simple<long>(RequestType.BitCount, [.. args]);
     }
 
-    public static Cmd<long, long> BitPositionAsync(ValkeyKey key, bool bit, long start, long end, StringIndexType indexType)
+    // Overload for StringIndexType (SER compatibility)
+    public static Cmd<long, long> BitCountAsync(ValkeyKey key, long start, long end, StringIndexType indexType)
+        => BitCountAsync(key, start, end, indexType.ToBitmapIndexType());
```
Comment (currantw) [inline sources/Valkey.Glide/Internals/Request.BitmapCommands.cs]: I think it probably makes sense to isolate the SER-only types and SER-only classes if possible. Can we do the conversion in `Database` and then use the `BitmapIndexType` version instead?

File: sources/Valkey.Glide/Internals/Request.BitmapCommands.cs [hunk 3]
```diff
@@ -14,48 +14,58 @@ public static Cmd<long, bool> GetBitAsync(ValkeyKey key, long offset)
...
         return Simple<long>(RequestType.BitOp, [.. args]);
     }
 
-    public static Cmd<object[], long[]> BitFieldAsync(ValkeyKey key, BitFieldOptions.IBitFieldSubCommand[] subCommands)
+    // Overload for two-key BITOP (SER compatibility)
+    public static Cmd<long, long> BitOpAsync(Bitwise operation, ValkeyKey destination, ValkeyKey first, ValkeyKey second)
+    {
+        GlideString[] args = [ValkeyLiterals.Get(operation).ToGlideString(), destination.ToGlideString(), first.ToGlideString(), second.ToGlideString()];
+        return Simple<long>(RequestType.BitOp, args);
+    }
```
Comment (currantw) [inline sources/Valkey.Glide/Internals/Request.BitmapCommands.cs]: Similarly, can we just route this through the existing `BitOpAsync` above?

File: sources/Valkey.Glide/Client/IBaseClient.BitmapCommands.cs [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    /// Return the position of the first bit set to 1 or 0 in a string.
+    /// </summary>
+    /// <seealso href="https://valkey.io/commands/bitpos"/>
+    /// <param name="key">The key of the string.</param>
+    /// <param name="bit">The bit value to search for (true for 1, false for 0).</param>
+    /// <param name="start">The start offset. Default is 0.</param>
+    /// <param name="end">The end offset. Default is -1 (end of string).</param>
+    /// <param name="indexType">The index type (bit or byte). Default is <see cref="BitmapIndexType.Byte"/>.</param>
+    /// <returns>The position of the first bit with the specified value, or -1 if not found.</returns>
+    Task<long> BitPosAsync(ValkeyKey key, bool bit, long start = 0, long end = -1, BitmapIndexType indexType = BitmapIndexType.Byte);
```
Comment (currantw) [inline sources/Valkey.Glide/Client/IBaseClient.BitmapCommands.cs]: How are you thinking about the naming? I've been tending to go with the more verbose option (e.g. `BitPositionAsync`, `BitOperationAsync`). This _seems_ to not be the approach taken by the other clients – they tend to hew quite closely to the actual Valkey command name. My feeling is that the slightly more verbose versions are better for consistency with SER and generally more common in C#, but happy either way

File: sources/Valkey.Glide/abstract_Enums/BitmapIndexType.cs [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    /// <summary>
+    /// Converts a <see cref="StringIndexType"/> to <see cref="BitmapIndexType"/>.
+    /// </summary>
+    internal static BitmapIndexType ToBitmapIndexType(this StringIndexType indexType) => indexType switch
+    {
+        StringIndexType.Bit => BitmapIndexType.Bit,
+        StringIndexType.Byte => BitmapIndexType.Byte,
+        _ => throw new ArgumentOutOfRangeException(nameof(indexType)),
+    };
+}
```
Comment (currantw) [inline sources/Valkey.Glide/abstract_Enums/BitmapIndexType.cs]: Related to other comments. But perhaps these are better segregated elsewhere to separate SER, GLIDE command, and GLIDE internal logic a bit? 🤔

File: sources/Valkey.Glide/Client/IBaseClient.BitmapCommands.cs [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    /// Reads or modifies the array of bits representing the string stored at key based on the specified subcommands.
+    /// </summary>
+    /// <seealso href="https://valkey.io/commands/bitfield"/>
+    /// <param name="key">The key of the string.</param>
+    /// <param name="subCommands">The subcommands to execute (GET, SET, INCRBY, OVERFLOW).</param>
+    /// <returns>
+    /// An array of results from the executed subcommands.
+    /// Null values indicate overflow when using OVERFLOW FAIL.
+    /// </returns>
+    Task<long?[]> BitFieldAsync(ValkeyKey key, IEnumerable<BitFieldOptions.IBitFieldSubCommand> subCommands);
```
Comment (currantw) [inline sources/Valkey.Glide/Client/IBaseClient.BitmapCommands.cs]: Do we want a one-command-only convenience option? Seems to be what we are (often) doing (SER also does it but not super consistently).

File: tests/Valkey.Glide.IntegrationTests/StackExchange/BitmapCommandsTests.cs [hunk 7]
```diff
@@ -0,0 +1,493 @@
...
+        Assert.Equal(glideValue.ToString(), serValue.ToString());
+    }
+
+    #endregion
+}
+
+/// <summary>
+/// Fixture class for <see cref="BitmapCommandsTests" />.
+/// </summary>
+public class BitmapCommandsFixture : IDisposable
```
Comment (currantw) [inline tests/Valkey.Glide.IntegrationTests/StackExchange/BitmapCommandsTests.cs]: I'm gonna add `IDatabaseAsync` test data to `TestConfiguration` in my PR, so we should be able to avoiding creating a separate test fixture for each test class in the future.

**Review by currantw — COMMENTED** [review]
Reviewed to `b33b34dd680f5d0ea0b8c560d887f1682383029f`

File: sources/Valkey.Glide/Abstract/Database.BitmapCommands.cs [hunk 8]
```diff
@@ -8,58 +8,49 @@ namespace Valkey.Glide;
...
-        return await StringSetBitAsync(key, offset, value);
+        return await SetBitAsync(key, offset, value);
     }
 
     /// <inheritdoc cref="IDatabaseAsync.StringBitCountAsync(ValkeyKey, long, long, StringIndexType, CommandFlags)"/>
     public async Task<long> StringBitCountAsync(ValkeyKey key, long start = 0, long end = -1, StringIndexType indexType = StringIndexType.Byte, CommandFlags flags = CommandFlags.None)
     {
         GuardClauses.ThrowIfCommandFlags(flags);
-        return await StringBitCountAsync(key, start, end, indexType);
+        return await BitCountAsync(key, start, end, indexType.ToBitmapIndexType());
```
[the remark this answers is not in the document]
Comment (alexr-bq) [reply sources/Valkey.Glide/Abstract/Database.BitmapCommands.cs]: I view this as the SER layer is a compatibility layer and it's fine if it references the glide layer, but not vice-versa.

File: sources/Valkey.Glide/Internals/Request.BitmapCommands.cs [hunk 9]
```diff
@@ -14,48 +14,58 @@ public static Cmd<long, bool> GetBitAsync(ValkeyKey key, long offset)
...
         => new(RequestType.SetBit, [key.ToGlideString(), offset.ToGlideString(), value.ToGlideString()], false, response => response != 0);
 
-    public static Cmd<long, long> BitCountAsync(ValkeyKey key, long start, long end, StringIndexType indexType)
+    public static Cmd<long, long> BitCountAsync(ValkeyKey key, long start, long end, BitmapIndexType indexType)
     {
         List<GlideString> args = [key.ToGlideString(), start.ToGlideString(), end.ToGlideString()];
-        if (indexType != StringIndexType.Byte)
+        if (indexType != BitmapIndexType.Byte)
         {
             args.Add(indexType.ToLiteral().ToGlideString());
```
[the remark this answers is not in the document]
Comment (currantw) [reply sources/Valkey.Glide/Internals/Request.BitmapCommands.cs]: Ya agreed. I think we might try to address some easy performance improvements next release.

File: sources/Valkey.Glide/abstract_Enums/BitmapIndexType.cs (hunk 5 above)
Comment (currantw) [reply sources/Valkey.Glide/abstract_Enums/BitmapIndexType.cs]: Great! I'll start moving things there too on my end.

**Review by currantw — APPROVED** [review]
Reviewed to `b33b34dd680f5d0ea0b8c560d887f1682383029f`

**Review by jeremyprime — APPROVED** [review]
